### ISSUE TITLE
Improve SW Code Types 8,B,C - move onto the next if match not found

### ISSUE THREAD

**Issue by XxUnkn0wnxX** [bucanero/apollo-lib#11]
@bucanero
can we have it so it does not fail when it does not find something & simply move onto the next segment of the code block.

```text
[i] Applying codes [1] to GOWRSAVE12...

[+] Applying code bucanero/apollo-ps4#1...
- Game Genie Code
- Applying [Key Items - 2147400000] to 'GOWRSAVE12'...
- Searching (len=8 count=1) ...
- ----- Search 8 bytes -----
- 000000: CB 8B 5C CF C1 7F 4E 15 
- ----- Search 8 bytes -----
- SEARCH PATTERN NOT FOUND
- ERROR!
```

Example:
```text
[Key Items - 2147400000]
80010008 CB8B5CCF
C17F4E15 00000000
28000008 7FFEB940
80010008 3832900B <-- say it fails here (Cant Find)
4A8A614B 00000000 <-- skip
28000008 7FFEB940 <-- skip
80010008 0F293CEC <-- resume here
4D3EDEEE 00000000
28000008 7FFEB940
80010008 28D9B3C5
D702057B 00000000
28000008 7FFEB940
```

by default SW just skips anything it can't find & moves down the code block & adjusts the save for any matches.
like for all resources codeblock there would be multiple searches one after the other. the player may not have all items for the code to be fully effective. However, it still should match anything it can find and adjust accordingly.

**Comment by bucanero** [comment]
to properly reproduce the issue and test a solution, can you share:
- the `.savepatch` file with the codes to apply
- the data file you're trying to modify (`GOWRSAVE12`)

**Comment by XxUnkn0wnxX** [comment]
[Patch_Save.zip](https://github.com/bucanero/apollo-lib/files/10203364/Patch_Save.zip)
Patch + Save File..

**Comment by jrson83** [comment]
Just to add I noticed the same issue:

The most codes have a description `Must Have 1 in inventory or Storage`
Let's take the quick code for `Max Blood Rock` as example:

```
[Max Blood Rock (Must Have 1 in inventory or Storage)]
80010004 D60B0040
18000004 000003E7
80020004 D60B0040
18000004 000003E7
```

It will search for the Blood Rock hex-id `D60B0040` and apply the code. It will do that 2x, one time for inventory and one time for storage.

If the savegame has the item in both inventory & storage, the quick code patch will apply with no problems.

But when we have the item only in one, either inventory or storage, the first patch will apply, but since the second search will not find another code, it will throw and error `SEARCH PATTERN NOT FOUND`.

The output when testing the codes with library patcher cli is `Patching completed: 1 codes applied`, but it actually does not save the file after applying the patch.

So basically: when the item we want to apply a patch, is only either in inventory or storage, no patch gets applied at all, cause when an error occures it does not save the first applied patch to the file.

I can't tell how it works using Save Wizard.

> by default SW just skips anything it can't find & moves down the code block & adjusts the save for any matches.

If this is the case the quick codes are OKAY, since the first code would be applied and also **saved** to the file.

**Comment by bucanero** [comment]
one question, it would be enough to assume that if a search failed (code type `8xxxxx xxxxxx`), the process should skip and continue to the next search code , `8xxxxx xxxx`?

**Comment by jrson83** [comment]
> one question, would it be enough to assume that if a search failed (code type `8xxxxx xxxxxx`), the process should skip and continue to the next search code line, `8xxxxx xxxx`?

Yeah, I think for the most cases that should work.

**Comment by bucanero** [comment]
@XxUnkn0wnxX @jrson83 , could you test this build that implements a search skip until it finds the next search block?
I tested with the GOWR sample provided, but getting more testing would be great before adding it.

Win64: https://github.com/bucanero/apollo-lib/suites/11158917781/artifacts/569632980
Win32: https://github.com/bucanero/apollo-lib/suites/11158917781/artifacts/569632979

if you need linux or mac binaries, let me know.

**Comment by jrson83** [comment]
> @XxUnkn0wnxX @jrson83 , could you test this build that implements a search skip until it finds the next search block? I tested with the GOWR sample provided, but getting more testing would be great before adding it.
> 
> Win64: https://github.com/bucanero/apollo-lib/suites/11158917781/artifacts/569632980 Win32: https://github.com/bucanero/apollo-lib/suites/11158917781/artifacts/569632979
> 
> if you need linux or mac binaries, let me know.

I have tested both versions on windows 11. I first applied the `[Max Blood Stone Shard (Must Have 1 in inventory or Storage)]` patch while having a Blood Stone Shard in both inventory and storage. It worked without any issues:

```bash
./patcher userdata0000.savepatch 1 userdata0000

Apollo cheat patcher v0.3.0 - (c) 2022 by Bucanero

[i] Applying codes [1] to userdata0000...

[+] Applying code #1...
- Game Genie Code
- Applying [Max Blood Stone Shard (Must Have 1 in inventory or Storage)] to 'userdata0000'...
- Searching (len=4 count=1) ...
- ----- Search 4 bytes -----
- 000000: B8 0B 00 40
- ----- Search 4 bytes -----
- Search pointer = 109448 (0x1AB88)
- Wrote 2 bytes (03E7) to 0x1AB8C
- Searching (len=4 count=2) ...
- ----- Search 4 bytes -----
- 000000: B8 0B 00 40
- ----- Search 4 bytes -----
- Search pointer = 137988 (0x21B04)
- Wrote 2 bytes (03E7) to 0x21B08
- OK

Patching completed: 1 codes applied
```

Then I tried `[Max Blood Rock (Must Have 1 in inventory or Storage)]` while having a Blood Rock only inside my inventory. It still throws the same error and does not save any changes to the file, so still does not apply the first matched patch.

```bash
./patcher userdata0000.savepatch 1 userdata0000

Apollo cheat patcher v0.3.0 - (c) 2022 by Bucanero

[i] Applying codes [1] to userdata0000...

[+] Applying code #1...
- Game Genie Code
- Applying [Max Blood Rock (Must Have 1 in inventory or Storage)] to 'userdata0000'...
- Searching (len=4 count=1) ...
- ----- Search 4 bytes -----
- 000000: D6 0B 00 40
- ----- Search 4 bytes -----
- Search pointer = 113992 (0x1BD48)
- Wrote 2 bytes (03E7) to 0x1BD4C
- Searching (len=4 count=2) ...
- ----- Search 4 bytes -----
- 000000: D6 0B 00 40
- ----- Search 4 bytes -----
- SEARCH PATTERN NOT FOUND
- ERROR!

Patching completed: 1 codes applied
```

**Comment by bucanero** [comment]
@jrson83 thanks a lot for testing and sharing your results 👍 

I'll use your unpatched save data file, and test with your `Max blood rock` code:
```
80010004 D60B0040
18000004 000003E7
80020004 D60B0040
18000004 000003E7
```

I probably missed some edge case.

[save_test.zip](https://github.com/bucanero/apollo-lib/files/10817965/save_test.zip)

**Comment by bucanero** [comment]
@jrson83 , could you test out this new build? Now it should work in both cases, skipping until the following search code or saving the current changes if nothing else is found.

win32: https://github.com/bucanero/apollo-lib/suites/11189267309/artifacts/571861875
win64: https://github.com/bucanero/apollo-lib/suites/11189267309/artifacts/571861878

**Comment by jrson83** [comment]
> @jrson83 , could you test out this new build? Now it should work in both cases, skipping until the following search code or saving the current changes if nothing else is found.
> 
> win32: https://github.com/bucanero/apollo-lib/suites/11189267309/artifacts/571861875 win64: https://github.com/bucanero/apollo-lib/suites/11189267309/artifacts/571861878

@bucanero I tested again with both versions and it is now writing & saving the first bytes search, like it should. Awesome, thank you!

**Comment by bucanero** [comment]
Thanks for testing out @jrson83 👍  I'll include this fix in the next version release, once I add a few other small changes.
